### ISSUE TITLE
compiler 1.28.0 internal errors: it first expressed as a segfault

### ISSUE THREAD

**Issue by jlbyrne-hpe** [chapel-lang/chapel#21058]
### Summary of Problem

Currently :
foo.chpl:24: internal error: RES-LOW-ORS-1987 chpl version 1.28.0
If first expressed itself as a segv  being caught at utils/misc.c:948, but as I reduced the test case the expression changed.  Seems to be a use after free problem. 

### Steps to Reproduce

**Source Code:**

```chapel
module Foo {
    use PrivateDist;

    class foo {
    }

    class foo_context {
    }

    class FOOPerLoc  {
        var groupName: string;
        var fooInst: unmanaged foo? = nil;
        var fooContexts: [0..<here.maxTaskPar] unmanaged foo_context? = nil;
    }

    class FOOGlobal {
        var fooPerLoc: [PrivateSpace] unmanaged FOOPerLoc? = nil;
    }

    class FOOStore {
        proc type fooInit(groupName: string = "ignored") throws {
            var fooGlobal = new shared FOOGlobal();
            try {
                forall f in fooGlobal.fooPerLoc with (in groupName) {
                    f = new unmanaged FOOPerLoc();
                    f!.groupName = groupName;
                    f!.fooInst = new unmanaged foo();
                    // f!.fooInst!.foo_initialize(f!.groupName);
                    forall c in f!.fooContexts {
                        c = new unmanaged foo_context();
                        // c!.foo_ctx = f!.fooInst!.foo_context_open();
                    }
                }
            } catch e {
                throw e;
            }
        }
    }

    proc main() {
        try! {
            FOOStore.fooInit();
        }
	return 0;
    }
}
```

**Compile command:**

chpl -o foo foo.chpl

**Execution command:**

**Associated Future Test(s):**

### Configuration Information

- Output of `chpl --version`:
chpl version 1.28.0
  built with LLVM version 14.0.6
- Output of `$CHPL_HOME/util/printchplenv --anonymize`:
CHPL_TARGET_PLATFORM: hpe-cray-ex
CHPL_TARGET_COMPILER: clang *
CHPL_TARGET_ARCH: x86_64
CHPL_TARGET_CPU: unknown
CHPL_LOCALE_MODEL: flat
CHPL_COMM: none *
CHPL_TASKS: qthreads *
CHPL_LAUNCHER: none *
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc *
CHPL_ATOMICS: cstdlib
CHPL_GMP: bundled
CHPL_HWLOC: bundled
CHPL_RE2: bundled *
CHPL_LLVM: system *
CHPL_AUX_FILESYS: none
- Back-end compiler and version, e.g. `gcc --version` or `clang --version`:
clang version 14.0.6
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /opt/foo/llvm/14.0.6/bin
- (For Cray systems only) Output of `module list`:
Currently Loaded Modulefiles:
> 1) craype-x86-milan         4) perftools-base/22.09.0   7) cray-dsmml/0.2.2        10) cray-openshmemx/11.5.6  13) gt_llvm/14.0.6
> 2) libfabric/1.15.2.0       5) cce/14.0.4               8) PrgEnv-cray/8.3.3       11) cray-python/3.9.13.1
> 3) craype-network-ofi       6) craype/2.7.17            9) gcc/11.2.0              12) gt_launcher/4.0.7

**Comment by bradcray** [comment]
This passes valgrind compilation cleanly, so I don't think it's a use-after-free.  But that also makes me worry that the segfault you saw was something else.

I'm finding if I comment out the `try...catch` block around the `forall`, it compiles without complaint.

@vasslitvinov : This looks thick within code you've written if you can take a look.
@jlbyrne-hpe : Vass is ostensibly on vacation (though he keeps popping up, so...).  I'd be curious whether, if you comment out the try...catch blocks in your original code whether that makes the issue go away (which is not to say that's a reasonable thing to do, but it'd still be an interesting data point).

**Comment by jlbyrne-hpe** [comment]
@bradcray: Yes, the issue does go away and because the actual errors I'm catching will probably occur on all iterations  or none, I worked around the compilation problem by breaking out the first iteration in a try-catch and leaving rest of the loops out of a try-catch block.

**Comment by jlbyrne-hpe** [comment]
I just ran into this again in 1.30.0 and I'd forgotten it until I searched for PrivateDist in the issues. Have a somewhat smaller testcase.
```
module Foo2 {
    use PrivateDist;
    use List;

    record Bar {
        var value: int;
    }

    var perLocLists: [PrivateSpace] list(Bar);

    proc main() {
        try! {
            forall l in perLocLists {
                forall b in l {
                    b.value = 1;
                }
            }
        }
        return 0;
    }
}
chpl -o Foo2 Foo2.chpl
Foo2.chpl:13: internal error: RES-LOW-ORS-1987 chpl version 1.30.0
```

**Comment by bradcray** [comment]
@vasslitvinov : I suspect the ping above got lost in your post-vacation catch-up.  Would you be able to take a look and see whether addressing this would be low-hanging fruit?  It's hitting an assertion-style INT_FATAL that you touched last.

**Comment by jlbyrne-hpe** [comment]
When Brad suggesting taking out the try!, I did just that and the problem went away and I moved on. But today I had a thought I should have had before: moving the problem loop into a procedure and calling that from the try! compiles and runs without obvious problems and adding a throw to test that seemed to work as well.  (With 1.30.0)

**Comment by vasslitvinov** [comment]
Here are the results of my investigation last week.

I have a one-liner that leads to correct execution of the Foo2 test above, https://github.com/chapel-lang/chapel/issues/21058#issuecomment-1494990477 . This change does not help, however, when the follower iterator invoked in the inner forall loop (or the outer loop?) has cleanup actions across its yield statement(s).

<details>

```diff
--- compiler/resolution/lowerIterators.cpp
+++ compiler/resolution/lowerIterators.cpp
@@ -2034,6 +2034,8 @@
     if (tail->gotoTag == GOTO_RETURN)
       tail->remove();

+  if (breakBlock->body.empty()) return; // nothing to add
+
   addIteratorBreakBlocks(loopRef, IC, loopBody, breakBlock);
 }

```

</details>

For debugging, I used a variant of the Foo2 test that has small iterators.

<details>

```chpl
proc main() {
  try! {
    forall l in these2() {
      for b in these3() { // or forall
        writeln(b);
      }
    }
  }
  return 0;
}

iter these2() {
  yield 1;
}

iter these2(param tag: iterKind) where tag == iterKind.leader {
  yield 3;
}

iter these2(param tag: iterKind, followThis) where tag == iterKind.follower {
  defer { writeln(123412341234); }  // probably not needed
  yield 5;
}

iter these3() {
  defer { writeln(432143214321); }  // needed if the "one-liner" is applied
  yield 11;
}

iter these3(param tag: iterKind) {
  coforall 1..2 do
    yield 13;
}
```

</details>

If, in my Foo2 variant above, I annotate the parallel iterator `these3` with `throws`, the compiler fails an assertion. I did not investigate.

Given that the assertion failure in the original user code occurs in `ibbInsertPoint()` in lowerIterators.cpp, I thought: "why not just put the IBB right before the goto `gt` ? A1: this does not work in recursive iterators. A2: in the case of nested loops when both iterators (i.e., for the outer and the inner loops) have cleanup actions, those cleanup actions are executed in a wrong order. I have a slimmed-down version of a test that illustrates this.

<details>

```chpl
// extract from
//  test/functions/iterators/vass/defer-in-iterators/defer-in-iterators.chpl

config const doReturn = true;

record RRR {
  const jj:int;
  proc init(arg:int) { jj=arg; complete(); writeln("RRR.init { ", jj); }
  proc deinit() { writeln("RRR.fini } ", jj); }
}

iter myIter1() {
  writeln("start myIter1");

  defer { writeln("myIter1 defer 300"); }
  var loc1a = new RRR(300);
  yield new RRR(33);
/* defer actions for myIter1 when breaking out from a loop over it:
RRR.fini } 33
RRR.fini } 300
myIter1 defer 300
*/

  writeln("myIter1 exit");
}

iter myIter2() {
  defer { writeln("myIter2 defer 500"); }
  var loc2a = new RRR(500);

  for r2nz in myIter1() {
    writeln("myIter2 ", r2nz.jj);
    if doReturn  {
      writeln("myIter2 returning");
      return;
/* defer actions for myIter2 at this point:
those for myIter1
RRR.fini } 500
myIter2 defer 500
*/
    }
    yield new RRR(1000 + r2nz.jj);
  }

  writeln("myIter2 exit");
}

proc main {
  writeln("test2nz start");

  for r2nz in myIter2() {
    writeln("test2nz ", r2nz.jj);
  }

  writeln("test2nz finish");
}
```

</details>

Next steps: figure out where the IBB needs to go and write code to find that spot in the AST.
